text/3617-precise-capturing.md

```diff
@@ -0,0 +1,670 @@
+
+### `impl move<'t, T> Trait`
+
+We could use the existing `move` keyword, however the word "move" is semantically worse.  In Rust, we already *use* generic parameters in types, but we don't *move* any generic parameters.  We move only *values*, so this could be confusing.  The word "use" is better.
+
+### `impl k#via<'t, T> Trait`
+
+We could use a new short keyword such as `via`.  This has the number 1 and 2 drawbacks of `k#captures` mentioned above.  As with `move`, it also seems a semantically worse word.  With `use<..>`, we can explain that it means the opaque type *uses* the listed generic parameters.  In contrast, it's not clear how we could explain the word "via" in this context.
+
+### Using parenthesis or square brackets
+
+We could say `use('t, T)` or `use['t, T]`.  However, in Rust today, generic parameters always fall within angle brackets, even when being applied to a type.  Doing something different here could feel inconsistent and doesn't seem warranted.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Opting out of captures
+
+There will plausibly be cases where we want to capture many generic parameters and not capture only smaller number.  It could be convenient if there were a way to express this without listing out all of the in-scope type parameters except the ones not being captured.
+
+The way we would approach this with the `use<..>` syntax is to add some syntax that means "fill in all in-scope generic parameters", then add syntax to remove certain generic parameters from the list.  E.g.:
+
+```rust
+fn foo<'a, A, B, C, D>(
+    _: &'a A, b: B, c: C, d: D,
+) -> impl use<.., !'a, !A> Sized {}
+//   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
+//   ^ Captures `B`, `C`, and `D` but not `'a` or `A`.
+```
+
+Here, the `..` means to include all in-scope generic parameters and `!` means to exclude a particular generic parameter even if previously included.
+
+We leave this to future work.
+
+## Explicit capturing for closure-like blocks
+
+Closures and closure-like blocks (e.g. `async`, `gen`, `async gen`, `async` closures, `gen` closures, `async gen` closures, etc.) return opaque types that capture both *values* and *generic parameters* from the outer scope.
+
+### Specifying captured generics for closures-like blocks
+
+The capturing of outer generics in closure-like blocks can lead to overcapturing, as in [#65442][].  Consider:
+
+```rust
+trait Trait {
+    type Ty;
+    fn define<T>(_: T) -> Self::Ty;
+}
+
+impl Trait for () {
+    type Ty = impl Fn();
+    fn define<T>(_: T) -> Self::Ty {
+        || ()
+    //~^ ERROR type parameter `T` is part of concrete type but not
+    //~|       used in parameter list for the `impl Trait` type alias
+    }
+}
+```
+
+Here, the opaque type of the closure is capturing `T`.  We may want a way to specify which outer generic parameters are captured by closure-like blocks.  We could apply the `use<..>` syntax to closure-like blocks to solve this, e.g.:
```

Comment:
Why can't this "just work"? Shouldn't the compiler be able to figure out that the closure is not using `T`? There should be no semver hazard, because to return the closure you need RPIT, at which point you as API designer have the opportunity to specify the captures you commit to in public API.